classes/linked_list.py

```diff
@@ -0,0 +1,288 @@
+    * given index, removing data, removing an index, getting the data at a given index, and printing the data in the
+    * linked list.
+
+    The LinkedList class has the following attributes:
+    - `head`: A reference to the first node in the linked list.
+    - `tail`: A reference to the last node in the linked list.
+    - `size`: The number of elements in the linked list.
+
+    The LinkedList class has the following methods:
+    - `get_size`: Returns the size of the linked list.
+    - `append`: Appends data to the end of the linked list.
+    - `insert`: Inserts data at a given index in the linked list.
+    - `remove_data`: Removes the first appearance of data from the linked list.
+    - `remove_index`: Removes the specified index from the linked list.
+    - `get_data_at_index`: Returns the data at the given index.
+    - `print_ll_data`: Prints the data of the current linked list.
+
+    Example usage:
+    ```
+    # Create a new linked list
+    linked_list = LinkedList()
+
+    # Append data to the linked list
+    linked_list.append(1)
+    linked_list.append(2)
+    linked_list.append(3)
+
+    # Insert data at a given index in the linked list
+    linked_list.insert(0, 0)
+    linked_list.insert(4, 4)
+
+    # Remove data from the linked list
+    linked_list.remove_data(2)
+
+    # Remove an index from the linked list
+    linked_list.remove_index(0)
+
+    # Get the data at a given index in the linked list
+    data = linked_list.get_data_at_index(0)
+
+    # Print the data in the linked list
+    linked_list.print_ll_data()
+    """
+
+    def __init__(self):
+        self.head = None
+        self.tail = None
+        self.size = 0
+
+    def get_size(self) -> int:
+        """
+        Returns the size of the linked list.
+        :return: Number of elements in the linked list.
+        """
+        return self.size
+
+    def append(self, data: Any) -> None:
+        """
+        Appends data to the end of the linked list.
+        :param data: Data to be appended.
+        """
+        node = Node(data)
+        if self.size == 0:
+            self.head = node
+            self.tail = node
+        else:
+            self.tail.set_next(node)
+            self.tail = node
+        self.size += 1
+
+    def insert(self, data: Any, index: int) -> None:
+        """
+        Inserts data at a given index in the linked list.
+        :param data: Data to be inserted.
+        :param index: Index between 0 and size of the linked list.
+        """
+        if index == self.size:
+            self.append(data)
+            return
+        node = Node(data)
+        if index == 0:
+            node.set_next(self.head)
+            self.head = node
+            self.size += 1
+        elif 0 < index < self.size:
+            iter_node = self.head
+            for _ in range(0, index - 1):
+                iter_node = iter_node.get_next()
```

Comment:
I do not understand what is done here. you update the iter_node in the for loop and then you set the next node, right? but I do not understand node.set_next(iter_node.get_next())and iter_node.set_next(node)

Reply:
![image](https://github.com/SaraSalvador1991/CS-Seminars/assets/84495726/5e3c6504-8507-46e7-abdb-2ffa290a1cab)
I refer to the example image. If you want to insert the `node with value 15`, in the for loop you iterate through the linked list with the iter_node until you reach the `node with value 91`. The `node with value 23` is the next node of iter_node. With node.set_next(iter_node.get_next()) you set the `node with value 23` as the next node of the `node with value 15` (right red arrow in the image). With iter_node.set_next(node) you set the `node with value 15` as the next node of the `node with value 91` (left red arrow).

Reply:
thank you David for the explanation and the scheme, it's clearer now :)